File: components/go-api/logger.go
package main

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

type LogLevel string

const (
	LogLevelInfo  LogLevel = "info"
	LogLevelWarn  LogLevel = "warn"
	LogLevelError LogLevel = "error"
	LogLevelDebug LogLevel = "debug"
)

type LogInput struct {
	Action  string
	Message string
	Data    map[string]any
}

type logFormat struct {
	Timestamp string         `json:"timestamp"`
	Level     LogLevel       `json:"level"`
	Action    string         `json:"action"`
	Message   string         `json:"message"`
	Data      map[string]any `json:"data"`
}

type Logger struct{}

var (
	loggerInstance *Logger
	loggerOnce     sync.Once
)

func GetLogger() *Logger {
	loggerOnce.Do(func() {
		loggerInstance = &Logger{}
	})
	return loggerInstance
}

func (l *Logger) log(level LogLevel, input LogInput) {
	if input.Data == nil {
		input.Data = map[string]any{}
	}

	entry, err := json.Marshal(logFormat{
		Timestamp: time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
		Level:     level,
		Action:    input.Action,
		Message:   input.Message,
		Data:      input.Data,
	})
	if err != nil {
		return
	}

	fmt.Println(string(entry))
}

func (l *Logger) Info(input LogInput) {
	l.log(LogLevelInfo, input)
}

func (l *Logger) Warn(input LogInput) {
	l.log(LogLevelWarn, input)
}

func (l *Logger) Error(input LogInput) {
	l.log(LogLevelError, input)
}

func (l *Logger) Debug(input LogInput) {
	l.log(LogLevelDebug, input)
}

File: components/nodejs-api/src/logger.ts
enum LogLevel {
  INFO = "info",
  WARN = "warn",
  ERROR = "error",
  DEBUG = "debug",
}

type LogFormat = {
  timestamp: string;
  level: LogLevel;
  action: string;
  message: string;
  data: Record<string, unknown>;
};

export class Logger {
  private static instance: Logger;

  private constructor() {}

  public static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  private log(data: Omit<LogFormat, "timestamp">): void {
    console.log(
      JSON.stringify({
        timestamp: new Date().toISOString(),
        level: data.level,
        action: data.action,
        message: data.message,
        data: data.data,
      }),
    );
  }

  public info(data: Omit<LogFormat, "timestamp" | "level">): void {
    this.log({ ...data, level: LogLevel.INFO });
  }

  public warn(data: Omit<LogFormat, "timestamp" | "level">): void {
    this.log({ ...data, level: LogLevel.WARN });
  }

  public error(data: Omit<LogFormat, "timestamp" | "level">): void {
    this.log({ ...data, level: LogLevel.ERROR });
  }

  public debug(data: Omit<LogFormat, "timestamp" | "level">): void {
    this.log({ ...data, level: LogLevel.DEBUG });
  }
}
